scipy/spatial/transform/rotation.py

```diff
@@ -629,3 +656,17 @@ def inv(self):
         if self._single:
             quat = quat[0]
         return self.__class__(quat, normalized=True)
+
+    def __mul__(self, other):
+        p = self._quat
+        q = other._quat
+        q_mat = _rep_mat(q)
+
+        # einsum handles Nx1, NxN, 1xN for us
+        # raises ValueError if broadcasting is not possible
+        result = np.einsum('...ij,...ijk->...ik', p, q_mat)
```

Comment:
Given that we want to expand the object with the smaller number of rotations, I think it's best to handle each case `Nx1`, `NxN` and `1xN` separately. I'll make the changes and push.

Reply:
What gain is there by doing that? I don't think the 3x gain will be worth the additional code complexity.

Reply:
In the case of `Nx1` and `NxN` compositions, we have to convert the second object to a matrix and multiply.

For the case of `1xN` composition, we should probably convert the first object to a matrix (converting the second one seems wasteful). Since the order is now reversed, this composition will involved transposing the matrices before multiplication. This case needs to be handled differently anyway.

If we're aggressively trying to stick to one approach, then we could go with simple quaternion composition across the board. It's slower, but there are no cases attached to that.

Reply:
> It's slower, but there are no cases attached to that.

See my previous email about maintainability, numerical stability, and optimization -- I'd rather get clean and simple, more maintainable code working and merged, unless it's obvious that for expected realistic use cases there is a real benefit of adding code complexity. So if the quat composition isn't horribly slower (> 10x?) it seems fine to use it here. We can always optimize later when people hit limiting use cases (which might be never!).

Perhaps cleaner than this code would involve:

1. Allocating the output array based on input shape
2. Computing each output quat in a separate statement involving `[..., 0]`-style indexing.

Then I think you won't even need the `_single` check at the end because NumPy broadcasting will take care of this for you.

FWIW, if this is what you mean by "direct composition", I'm actually surprised that it's slower than what you wrote here. Both the `np.dot`-style operation and the "direct" method can make use of SIMD instructions for speed.

Reply:
Actually I'm not convinced the idea I had will work, so feel free to ignore it. But I am curious what the alternative code would look like, I imagine it would be simpler (and the `_rep_mat` would go away?)

Reply:
I've already uploaded the relevant benchmarks and code to the wiki [here](https://github.com/adbugger/scipy/wiki/Benchmarking-experiments-for-rotation-compositions).

For the `Nx1` case, the quaternion approach is ~4x slower until we reach astronomical numbers for `N`, where is becomes closer to ~8x slower.

For the `NxN` case, the quaternion approach is a little less then ~4x slower, but quickly starts outperforming the matrix approach at `~1e3` rotations.

Take a look and let me know.

Reply:
It's hard for me to understand how to approach optimization vs maintainability issues.

What Eric suggesting is to almost ignore all performance tricks. My problems with that it is unlikely to be changed afterwards, because other people won't get involved enough to figure it out. On the other hand, perhaps this "problem" is not that significant if it won't be obviously seen by others.

So if we adapt it, then I agree with Aditya:

> If we're aggressively trying to stick to one approach, then we could go with simple quaternion composition across the board. It's slower, but there are no cases attached to that.

I suggest to stick to this approach. It obviously has some advantages of being conceptually elegant and memory efficient.

-----------

As we are here, what about rotations applications? In this case we could again use quaternion formulation or get `dcm` by `as_dcm` and use dot product / einsum. In this case I'm even inclined to use `as_dcm` because it is already written and it is natural to apply transformation through matrix. (Also it will be quicker likely.).